added uns to misc slot migration in anndata2seurat

metakuni opened the pull request:
Addresses:
* https://github.com/chanzuckerberg/single-cell-data-portal/issues/1581

based on:
* https://github.com/chanzuckerberg/single-cell-curation/issues/149

pablo-gar reviewed (commented):
I'd recommend only saving the slots in `uns` that are in the schema. Datasets can come with many other slots and their types can be anything, I'm not sure how `reticulate::py_to_r` will handle certain python types. So I'd say transferring all of `uns` is fragile and can lead to unexpected errors.

Also, have you tried this out? It'd be great to make sure `reticulate::py_to_r` is working properly for each element of `uns`

metakuni commented:
> I'd recommend only saving the slots in `uns` that are in the schema. Datasets can come with many other slots and their types can be anything, I'm not sure how `reticulate::py_to_r` will handle certain python types. So I'd say transferring all of `uns` is fragile and can lead to unexpected errors.

Fixed. I've implemented filtering so that only the [`uns` keys defined in the schema](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/2.0.0/schema.md#uns-dataset-metadata) (`schema_version`, `title`, `X_normalization`, `batch_condition`, `default_embedding`, `X_approximate_distribution`) will be copied into the Seurat object's `misc` slot.

> Also, have you tried this out? It'd be great to make sure `reticulate::py_to_r` is working properly for each element of `uns`

I've run this code with a debugger using several AnnData test files and verified the contents of the `misc` slot in the generated Seurat files.

pablo-gar commented on R/functions.R:
```diff
@@ -184,6 +184,13 @@ loom2sce <- function(inFile, outFile = NULL, main_layer = NULL, main_layer_name
     return(var_df)
 }
 
+.uns2misc <- function(ad_pd) {
+    accepted_uns_keys <- c('schema_version', 'title', 'X_normalization', 'batch_condition', 'default_embedding', 'X_approximate_distribution')
+    uns_keys <- intersect(accepted_uns_keys, reticulate::py_to_r(ad_pd$uns_keys()))
+    misc <- sapply(uns_keys, function(x) reticulate::py_to_r(ad_pd$uns[x]), simplify = FALSE, USE.NAMES = TRUE)
+    return(misc)
+}
+
```
```suggestion
.uns2misc <- function(ad_pd, target_uns_keys = NULL) {
    uns_keys <- intersect(target_uns_keys, reticulate::py_to_r(ad_pd$uns_keys()))
    
    if (length (uns_keys) > 0) {
       misc <- sapply(uns_keys, function(x) reticulate::py_to_r(ad_pd$uns[x]), simplify = FALSE, USE.NAMES = TRUE)
    } else {
      misc <- list()
    }
    
    return(misc)
}
```

pablo-gar commented on R/functions.R:
```diff
@@ -296,6 +303,8 @@ anndata2seurat <- function(inFile, outFile = NULL, main_layer = 'counts', assay
         }
     }
 
+    srt@misc <- .uns2misc(ad)
```
```suggestion
    srt@misc <- .uns2misc(ad, target_uns_keys = target_uns_keys)
```

pablo-gar commented on R/functions.R:
```diff
@@ -184,6 +184,13 @@ loom2sce <- function(inFile, outFile = NULL, main_layer = NULL, main_layer_name
+    return(misc)
+}
+
 anndata2seurat <- function(inFile, outFile = NULL, main_layer = 'counts', assay = 'RNA', use_seurat = FALSE, lzf = FALSE) {
```
```suggestion
anndata2seurat <- function(inFile, outFile = NULL, main_layer = 'counts', assay = 'RNA', use_seurat = FALSE, lzf = FALSE, target_uns_keys = NULL) {
```

pablo-gar reviewed (commented):
I added the suggestion of passing the keys from uns to be transferred as an argument to `anndata2seurat`.

Then when we call `convertFormat` during the upload we can hardcode those keys.

I think this approach is better, because if and when we update the schema it will be cleaner and easier to update the keys in our end rather than having to update sceasy. And we are not the only user of sceasy so I don't think it's great to hardcode this for all users.

metakuni commented:
> I added the suggestion of passing the keys from uns to be transferred as an argument to `anndata2seurat`.
> 
> Then when we call `convertFormat` during the upload we can hardcode those keys.
> 
> I think this approach is better, because if and when we update the schema it will be cleaner and easier to update the keys in our end rather than having to update sceasy. And we are not the only user of sceasy so I don't think it's great to hardcode this for all users.

That's definitely better. Thanks for the suggestion!

metakuni commented:
Created a PR on `cellgeni/sceasy`:
* https://github.com/cellgeni/sceasy/pull/42
